### ISSUE TITLE
About pruning model on multi GPUs

### ISSUE THREAD

**Issue by qingzhouzhen** [microsoft/nni#5205]
is there any example about pruning model on multi GPUs? I had tried prune mode with multi GPU, but it failed(forget to collect log), then I prune model with one GPU and save model to disk, after that, I load the model and finetune the model with multi GPUs.

**Comment by J-shang** [comment]
hello @qingzhouzhen , which pruner do you use? you could show your code if convenience, we are going to write some examples on multi-gpu recently.

**Comment by J-shang** [comment]
speedup will change the model, so if you mean multi-gpu is DDP, you should put the original model into DDP wrapper after speedup.

have you tried prune model with multi-gpu, then rewrap the model with DDP and finetune.

**Comment by XYAskWhy** [comment]
> 
What if we use _pruing while training_ methods, such as TaylorFO and Slimming?

**Comment by J-shang** [comment]
> > 
> 
> What if we use _pruing while training_ methods, such as TaylorFO and Slimming?

You could use pruner with DDP.
Additionally, we found that TaylorFO has bug in DDP (pruning is independently on each device), we will fix this in v3.0.
Slimming seems work well in DDP, you could directly prune model in DDP and generate masks.

If speedup is used, you should rewrap the model by DDP, this is because speedup will replace the layers in the original model to smaller ones, then the old DDP wrapper might get wrong.

The entire pipeline is like:
```
ddp_model = DistributedDataParallel(model, ...)
pruner = XXPruner(ddp_model, config_list, ...)
wrapped_model, masks = pruner.compress()

pruner._unwrap_model()
ModelSpeedup(ddp_model, dummy_input, masks).speedup_model()
new_ddp_model = DistributedDataParallel(ddp_model.module, ...)
```

**Comment by XYAskWhy** [comment]
Hey, after careful experiments, I found Slimming have similar problem with TaylorFO. If we prune while DDP training, we always end up with different pruned model on different GPU after speedup. For TaylorFO, the prune training process can't sync grads, thus leads to different weights on different GPU, and finally leads to different pruned model on different GPUs which can't be continued to further DDP finetuning. For slimming, normal Batchnorm in DDP doesn't sync parameter across GPUs, so it end up with different pruned model on different GPUs as well.

**Comment by J-shang** [comment]
> Hey, after careful experiments, I found Slimming have similar problem with TaylorFO. If we prune model while DDP training, we always end up with different pruned model on different GPU after speedup. For TaylorFO, the prune training process can't sync grad, thus leads to different weights on different GPU, and finally leads to different pruned model on different GPU which can't be continued to further DDP finetuning. For Slimming, normal Batchnorm op doesn't sync bn parameters(or should be called as buffers) across GPUs, and nni doesn't support SyncBatchnorm op, so it end up with different pruned model on different GPU as well. If we wrap the model again after speedup, model on rank0 will replace all other models on all other ranks and subsequent finetuning will be ok, but it won't solve the real problem because this way we are actually pruning in a single process on a single GPU.

Thanks for your feedback, it's helpful!
We are trying to fix DDP issues in https://github.com/microsoft/nni/pull/5217 , and it will merge in v3.0.

For `SlimPruner`, you could have a try to modify:
https://github.com/microsoft/nni/blob/c31d2574cb418acfb80c17bb2bd03531556325bd/nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py#L610

to

```
sub_shcema[SchemaOptional('op_types')] = ['BatchNorm2d', 'SyncBatchNorm']
```

FYI @Bonytu

**Comment by XYAskWhy** [comment]
I did what you suggested, and registered `replace_syncbatchnorm` func in compress_modules.py to make speedup work.
It turns out that pruned models on different GPUs are 'almost' identical, but a few layers are still not same. Which is quite puzzling since bn buffers across GPUs seems to be equal now. Did you used random number or random function in the slimming algorithm?

**Comment by J-shang** [comment]
@XYAskWhy could you show a demo to reproduce your issue? I think nni did not use random gen in slimming algorithm, but random is used in speedup for sparsity propagation.

**Comment by XYAskWhy** [comment]
OK,I will make a demo for you next Monday. Thank you and have a nice weekend.

**Comment by XYAskWhy** [comment]
Sorry I'm not authorized to share the exact code, but I can share the gist of it. I'm using nni in mmdetection to prune mask-rcnn-r50, what i did is to insert nni's TaylorFO/Slimming prune code
 in epoch_based_runner.py of mmcv. ( https://github.com/open-mmlab/mmcv/blob/abb600ad8d34c0d1bc88b01c4d2526ebdea1f8dc/mmcv/runner/epoch_based_runner.py#L118)

**Comment by J-shang** [comment]
Thank you, we will have a try.
